### PR TITLE
feat: add groth16

### PR DESCRIPTION
**Pull request by mattstam**
Adds Groth16 to the circuits artifacts generated, and supports generating and verifying proofs for it.

Notes:
* Does not implement a real trusted setup
* The circuit tar contains files for both plonk and groth16, so that provers that download it will be able to prove for either proof type.
* Instead of circuit artifacts stored in `~/.sp1/circuits/plonk_bn254/<version>/`, they are now stored in `~/.sp1/circuits/<version>/`, and the full output looks like:
```
-rw-rw-r-- 1 ubuntu ubuntu  249660953 Aug 16 20:05 constraints.json
-rw-rw-r-- 1 ubuntu ubuntu  889966358 Aug 16 20:12 groth16_circuit.bin
-rw-rw-r-- 1 ubuntu ubuntu 2691757795 Aug 16 20:13 groth16_pk.bin
-rw-rw-r-- 1 ubuntu ubuntu       2509 Aug 16 20:13 Groth16SP1Verifier.sol
-rw-rw-r-- 1 ubuntu ubuntu      31661 Aug 16 20:12 Groth16Verifier.sol
-rw-rw-r-- 1 ubuntu ubuntu        556 Aug 16 20:12 groth16_vk.bin
-rw-rw-r-- 1 ubuntu ubuntu    1021306 Aug 16 20:05 groth16_witness.json
-rw-rw-r-- 1 ubuntu ubuntu  828550683 Aug 16 19:55 plonk_circuit.bin
-rw-rw-r-- 1 ubuntu ubuntu 4295001768 Aug 16 19:57 plonk_pk.bin
-rw-rw-r-- 1 ubuntu ubuntu       2503 Aug 16 19:57 PlonkSP1Verifier.sol
-rw-rw-r-- 1 ubuntu ubuntu      58508 Aug 16 19:55 PlonkVerifier.sol
-rw-rw-r-- 1 ubuntu ubuntu      34368 Aug 16 19:55 plonk_vk.bin
-rw-rw-r-- 1 ubuntu ubuntu    1021739 Aug 16 19:10 plonk_witness.json
-rw-rw-r-- 1 ubuntu ubuntu 3225633956 Aug 16 19:14 srs.bin
-rw-rw-r-- 1 ubuntu ubuntu 2147517604 Aug 16 19:50 srs_lagrange.bin
```

### CODE REVIEW

File: book/developers/building-plonk-artifacts.md [hunk 1]
```diff
@@ -1,8 +1,8 @@
-# Building PLONK Artifacts
+# Building circuit Artifacts
```
Comment (jtguibas) [inline book/developers/building-plonk-artifacts.md:1]: nit: circuit -> Circuit

Comment (mattstam) [reply book/developers/building-plonk-artifacts.md]: fixed

File: crates/prover/src/types.rs [hunk 2]
```diff
@@ -144,19 +144,45 @@ pub type SP1ReducedProof = SP1ProofWithMetadata<SP1ReducedProofData>;
...
 
 #[derive(Serialize, Deserialize, Clone)]
 pub struct SP1PlonkBn254ProofData(pub PlonkBn254Proof);
 
 #[derive(Serialize, Deserialize, Clone)]
-pub struct SP1PlonkProofData(pub PlonkBn254Proof);
+pub struct SP1Groth16Bn254ProofData(pub Groth16Bn254Proof);
+
+#[derive(Serialize, Deserialize, Clone)]
+pub enum SP1ProofData {
```
[the remark this answers is not in the document]
Comment (jtguibas) [reply crates/prover/src/types.rs:166]: SP1ProofData is a bit too general.

File: crates/prover/src/types.rs [hunk 3]
```diff
@@ -144,19 +144,45 @@ pub type SP1ReducedProof = SP1ProofWithMetadata<SP1ReducedProofData>;
...
+    Plonk(PlonkBn254Proof),
+    Groth16(Groth16Bn254Proof),
+}
+
+impl SP1ProofData {
+    pub fn get_proof_system(&self) -> &str {
```
Comment (jtguibas) [inline crates/prover/src/types.rs:172]: nit: can you use an enum here

**Review by jtguibas — APPROVED** [review]

File: crates/prover/src/types.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattstam) [reply crates/prover/src/types.rs]: fixed

File: crates/prover/src/types.rs (hunk 3 above)
Comment (mattstam) [reply crates/prover/src/types.rs]: fixed

**Pull request merged**